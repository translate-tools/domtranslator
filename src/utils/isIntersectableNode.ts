export function isIntersectableNode(node: Element) {
	if (node.nodeName === 'OPTION') return false;

	return document.body.contains(node);
}
