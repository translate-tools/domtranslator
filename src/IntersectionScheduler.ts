import { DOMTranslationScheduler } from './types';
import { NodesIntersectionObserver } from './utils/NodesIntersectionObserver';

/**
 * Scheduler for DOM translation that trigger callbacks only for visible nodes.
 * If node is not visible, callback will be triggered once node will intersect viewport.
 */
export class IntersectionScheduler implements DOMTranslationScheduler {
	private readonly intersectionObserver;
	constructor(intersectionConfig?: IntersectionObserverInit) {
		this.intersectionObserver = new NodesIntersectionObserver(intersectionConfig);
	}

	public add(node: Node, callback: (node: Node) => void): void {
		this.intersectionObserver.observe(node, callback);
	}

	public remove(node: Node): void {
		this.intersectionObserver.unobserve(node);
	}
}
