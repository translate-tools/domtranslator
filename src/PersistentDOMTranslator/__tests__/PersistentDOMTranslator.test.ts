import { readFileSync } from 'fs';

import {
	awaitTranslation,
	containsRegex,
	endsWithRegex,
	startsWithRegex,
	TRANSLATION_SYMBOL,
	translator,
} from '../../__tests__/utils';
import { DOMTranslator } from '../../DOMTranslator';
import { IntersectionScheduler } from '../../IntersectionScheduler';
import { NodesTranslator } from '../../NodesTranslator';
import { createNodesFilter, NodesFilterOptions } from '../../utils/nodes';

import { PersistentDOMTranslator } from '..';

require('intersection-observer');

(IntersectionObserver.prototype as any).POLL_INTERVAL = 100;

const getElementText = (elm: Element | null) =>
	elm && elm.textContent ? elm.textContent.trim() : null;

const composeName = (...args: (string | boolean)[]) => args.filter(Boolean).join(' ');

const fillDocument = (text: string) => {
	document.write(text);
};

beforeEach(() => {
	document.body.innerHTML = '';
	vi.clearAllMocks();
});

describe('basic usage', () => {
	const sample = readFileSync(__dirname + '/sample.html', 'utf8');

	[true, false].forEach((lazyTranslate) => {
		const testName = composeName(
			'translate whole document',
			lazyTranslate && 'with lazyTranslate',
		);
		test(testName, async () => {
			fillDocument(sample);
			const parsedHTML = document.documentElement.outerHTML;

			// Translate document
			const persistentTranslator = new PersistentDOMTranslator(
				new DOMTranslator(new NodesTranslator(translator), {
					filter: createNodesFilter(),
					scheduler: lazyTranslate ? new IntersectionScheduler() : undefined,
				}),
			);
			persistentTranslator.translate(document.documentElement);

			await awaitTranslation();
			expect(document.documentElement.outerHTML).toMatchSnapshot();

			// Disable translation
			persistentTranslator.restore(document.documentElement);
			expect(document.documentElement.outerHTML).toBe(parsedHTML);
		});
	});
});

[true, false].forEach((isLazyTranslation) =>
	describe(
		'translation with consider translatable and ignored nodes' +
			(isLazyTranslation ? ' (lazy translation mode)' : ''),
		() => {
			const sample = readFileSync(__dirname + '/sample.html', 'utf8');

			const filterOptions = {
				attributesList: ['title', 'alt', 'placeholder', 'label', 'aria-label'],
				ignoredSelectors: [
					'meta',
					'link',
					'script',
					'noscript',
					'style',
					'code',
					'textarea',
				],
			} satisfies NodesFilterOptions;

			test('translate whole document', async () => {
				fillDocument(sample);
				const parsedHTML = document.documentElement.outerHTML;

				// Translate document
				const persistentTranslator = new PersistentDOMTranslator(
					new DOMTranslator(new NodesTranslator(translator), {
						filter: createNodesFilter(filterOptions),
						scheduler: isLazyTranslation
							? new IntersectionScheduler()
							: undefined,
					}),
				);

				persistentTranslator.translate(document.documentElement);

				await awaitTranslation();
				expect(document.documentElement.outerHTML).toMatchSnapshot();

				// Disable translation
				persistentTranslator.restore(document.documentElement);
				expect(document.documentElement.outerHTML).toBe(parsedHTML);
			});

			test('translate changed nodes', async () => {
				fillDocument(sample);

				// Translate document
				const persistentTranslator = new PersistentDOMTranslator(
					new DOMTranslator(new NodesTranslator(translator), {
						filter: createNodesFilter(filterOptions),
						scheduler: isLazyTranslation
							? new IntersectionScheduler()
							: undefined,
					}),
				);
				persistentTranslator.translate(document.documentElement);

				await awaitTranslation();

				const div1 = document.createElement('div');
				document.body.appendChild(div1);

				div1.innerHTML = 'Text 1';
				await awaitTranslation();
				expect(div1.innerHTML).toMatch(startsWithRegex(TRANSLATION_SYMBOL));

				div1.innerHTML = 'Text 2';
				await awaitTranslation();
				expect(div1.innerHTML).toMatch(startsWithRegex(TRANSLATION_SYMBOL));

				const elmA = document.querySelector('a');
				expect(elmA).not.toBeNull();

				if (elmA !== null) {
					elmA.innerHTML = 'changed link text';
					elmA.setAttribute('title', 'changed title');
					elmA.setAttribute('href', 'changed url');

					await awaitTranslation();
					expect(elmA.innerHTML).toMatch(startsWithRegex(TRANSLATION_SYMBOL));
					expect(elmA.innerHTML).toMatch(endsWithRegex('changed link text'));

					expect(elmA.getAttribute('title')).toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);
					expect(elmA.getAttribute('href')).not.toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);
				}

				// Disable translation
				persistentTranslator.restore(document.documentElement);
				expect(div1.innerHTML).not.toMatch(startsWithRegex(TRANSLATION_SYMBOL));

				if (elmA !== null) {
					expect(elmA.innerHTML).not.toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);
					expect(elmA.getAttribute('title')).not.toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);
				}
			});

			test('translate multiple nodes', async () => {
				fillDocument(sample);

				// Translate document
				const persistentTranslator = new PersistentDOMTranslator(
					new DOMTranslator(new NodesTranslator(translator), {
						filter: createNodesFilter(filterOptions),
						scheduler: isLazyTranslation
							? new IntersectionScheduler()
							: undefined,
					}),
				);

				const pElm = document.querySelector('p');
				const form = document.querySelector('form');
				const figure = document.querySelector('figure');

				if (!pElm || !form || !figure)
					throw new Error('Not found elements for test');

				persistentTranslator.translate(form);
				persistentTranslator.translate(figure);

				await awaitTranslation();

				expect(getElementText(pElm)).not.toContain(TRANSLATION_SYMBOL);
				expect(getElementText(figure)).toContain(TRANSLATION_SYMBOL);
				expect(getElementText(form)).toContain(TRANSLATION_SYMBOL);

				// Disable translation
				persistentTranslator.restore(form);
				expect(getElementText(form)).not.toContain(TRANSLATION_SYMBOL);
				expect(getElementText(figure)).toContain(TRANSLATION_SYMBOL);

				persistentTranslator.restore(figure);
				expect(getElementText(form)).not.toContain(TRANSLATION_SYMBOL);
				expect(getElementText(figure)).not.toContain(TRANSLATION_SYMBOL);

				// Enable translation back
				persistentTranslator.translate(form);
				persistentTranslator.translate(figure);
				await awaitTranslation();

				expect(getElementText(figure)).toContain(TRANSLATION_SYMBOL);
				expect(getElementText(form)).toContain(TRANSLATION_SYMBOL);

				// Disable translation for all elements
				persistentTranslator.restore(form);
				persistentTranslator.restore(figure);
				expect(getElementText(form)).not.toContain(TRANSLATION_SYMBOL);
				expect(getElementText(figure)).not.toContain(TRANSLATION_SYMBOL);
			});

			test('use custom nodes filter', async () => {
				fillDocument(sample);

				// Translate document
				const persistentTranslator = new PersistentDOMTranslator(
					new DOMTranslator(new NodesTranslator(translator), {
						filter: createNodesFilter({
							...filterOptions,
							ignoredSelectors: [
								...filterOptions.ignoredSelectors,
								'[translate="no"], .notranslate, [contenteditable], [contenteditable="true"]',
								'.custom-elements :checked',
							],
						}),
						scheduler: isLazyTranslation
							? new IntersectionScheduler()
							: undefined,
					}),
				);

				persistentTranslator.translate(document.documentElement);

				await awaitTranslation();

				['[contenteditable]', '.notranslate', '[translate="no"]'].forEach(
					(selector) => {
						const element = document.querySelector(selector);
						expect(element).toBeInstanceOf(Element);
						expect(getElementText(element)).not.toMatch(
							containsRegex(TRANSLATION_SYMBOL),
						);
					},
				);

				// Considered even pseudo classes
				expect(
					document
						.querySelector('.custom-elements [type="checkbox"]:checked')
						?.getAttribute('title'),
				).not.toMatch(containsRegex(TRANSLATION_SYMBOL));
				expect(
					document
						.querySelector('.custom-elements [type="checkbox"]:not(:checked)')
						?.getAttribute('title'),
				).toMatch(containsRegex(TRANSLATION_SYMBOL));

				expect(document.documentElement.outerHTML).toMatchSnapshot();

				// Disable translation
				persistentTranslator.restore(document.documentElement);
			});

			describe('Case #147', () => {
				test('Elements with empty text is translated when text is inserted', async () => {
					// We trying to catch translator on incorrect "optimization" here.
					// If translator see "empty" node and say "ok, i don't care about it anymore",
					// it will lead us to problems with translating of dynamic elements.
					// For example, chat in Google Meet may create element with empty text first,
					// and then fill it with text. This test case is emulates described scenario.

					document.body.innerHTML = `<div id="root"><div id="foo">foo</div><div id="bar"></div></div>`;

					const rootElement = document.querySelector('#root');
					const fooElement = document.querySelector('#foo');
					const barElement = document.querySelector('#bar');

					// All elements found
					expect(rootElement).toBeInstanceOf(HTMLElement);
					expect(fooElement).toBeInstanceOf(HTMLElement);
					expect(barElement).toBeInstanceOf(HTMLElement);

					const translatorSpy = vi.fn(translator);
					const persistentTranslator = new PersistentDOMTranslator(
						new DOMTranslator(new NodesTranslator(translatorSpy), {
							// All attributes is not for translation
							// Our intention is to skip translation for node with id `bar`
							filter: (node) => !(node instanceof Attr),
							scheduler: isLazyTranslation
								? new IntersectionScheduler()
								: undefined,
						}),
					);

					// Translation runs fine
					await expect(
						Promise.all([
							persistentTranslator.translate(document.documentElement),
							awaitTranslation(),
						]),
					).resolves.not.toThrow();

					// All attributes is not translated
					expect(rootElement?.id).toBe('root');
					expect(fooElement?.id).toBe('foo');
					expect(barElement?.id).toBe('bar');

					// Node with text is translated
					expect(fooElement?.textContent).toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);

					// Node with no text stay with empty text (have no translation)
					expect(barElement?.textContent).toBe('');

					// We add empty text node that must not be translated (because no content for translation)
					const textNode = new Text('');
					barElement?.append(textNode);

					await awaitTranslation();
					expect(barElement?.textContent).toBe('');

					// Now we change value of empty text node
					// This text must be translated
					textNode.nodeValue = 'Another content';
					await awaitTranslation();
					expect(barElement?.textContent).toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);

					persistentTranslator.restore(document.documentElement);
				});

				test('Ignored nodes may be updated with no errors and no translator calls', async () => {
					document.body.innerHTML = `<div id="root"><div id="foo">foo</div><div id="bar">bar</div><div id="baz">baz</div></div>`;

					const rootElement = document.querySelector('#root');
					const fooElement = document.querySelector('#foo');
					const barElement = document.querySelector('#bar');
					const bazElement = document.querySelector('#baz');

					// All elements found
					expect(rootElement).toBeInstanceOf(HTMLElement);
					expect(fooElement).toBeInstanceOf(HTMLElement);
					expect(barElement).toBeInstanceOf(HTMLElement);
					expect(bazElement).toBeInstanceOf(HTMLElement);

					const translatorSpy = vi.fn(translator);
					const persistentTranslator = new PersistentDOMTranslator(
						new DOMTranslator(new NodesTranslator(translatorSpy), {
							filter: (node) => {
								if ((barElement as HTMLElement).contains(node))
									return false;
								if (
									node instanceof Attr &&
									node.ownerElement === barElement
								)
									return false;

								return true;
							},
							scheduler: isLazyTranslation
								? new IntersectionScheduler()
								: undefined,
						}),
					);

					// Translation runs fine
					await expect(
						Promise.all([
							persistentTranslator.translate(rootElement as HTMLElement),
							awaitTranslation(),
						]),
					).resolves.not.toThrow();

					// Translated all nodes except ignored
					expect((rootElement as HTMLElement).id).toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);
					expect((fooElement as HTMLElement).textContent).toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);
					expect((fooElement as HTMLElement).id).toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);
					expect((bazElement as HTMLElement).textContent).toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);
					expect((bazElement as HTMLElement).id).toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);

					expect((barElement as HTMLElement).textContent).not.toMatch(
						startsWithRegex(TRANSLATION_SYMBOL),
					);

					expect(translatorSpy).toBeCalledTimes(5);

					// Update of ignored node causes no translation calls or any errors
					expect(barElement?.id).toBe('bar');
					expect(barElement?.childNodes[0].textContent).toBe('bar');
					(barElement as HTMLElement).id = 'another-bar-id';
					(barElement as HTMLElement).childNodes[0].nodeValue =
						'another bar text';

					await expect(awaitTranslation()).resolves.not.toThrow();
					expect(translatorSpy).toBeCalledTimes(5);
					expect(barElement?.id).toBe('another-bar-id');
					expect(barElement?.childNodes[0].textContent).toBe(
						'another bar text',
					);
				});
			});
		},
	),
);
