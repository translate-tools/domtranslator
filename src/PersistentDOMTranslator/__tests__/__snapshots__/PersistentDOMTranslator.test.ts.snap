// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic usage > translate whole document 1`] = `
"<html lang="en"><head>
	<meta charset="UTF-8">
	<title>***TRANSLATED***Demo page with challenges for DOM translator</title>
	<meta name="viewport" content="width=device-width, initial-scale=1.0" data-rh="true">
	<link rel="search" type="application/opensearchdescription+xml" title="Opensearch" href="/opensearch.xml">
	<link rel="preload" href="#file.js" as="script">
	<script>***TRANSLATED***console.log('console log: page start')</script>
</head>

<body>***TRANSLATED***
	Text with no container
	<div id="rootElement" class="main container">***TRANSLATED***
		Text in div<br>
		<p>***TRANSLATED***Text inside container <a href="#url" title="link title">***TRANSLATED***link text</a></p>

		<img title="image" alt="image alt text" data-text="data attribute that must not be translated">
		<pre>***TRANSLATED***Text with preserved formatting</pre>***TRANSLATED***

		Some code:
		<pre>			<code lang="javascript">***TRANSLATED***
				const name = "Jeff";
				console.log("Your name is " + name);
			</code>
		</pre>

		<figure>
			<blockquote cite="https://www.huxley.net/bnw/four.html">
				<p>***TRANSLATED***Words can be like X-rays, if you use them properly—they’ll go through anything. You read and you’re pierced.</p>
			</blockquote>
			<figcaption>***TRANSLATED***—Aldous Huxley, <cite>***TRANSLATED***Brave New World</cite></figcaption>
		</figure>

		<form method="post" style="display: flex; flex-flow: column;">
			<input type="text" name="text" value="text">
			<input type="password" name="text" value="password">
			<textarea>***TRANSLATED***long text in textarea</textarea>

			<label><input type="checkbox">***TRANSLATED*** checkbox <i>***TRANSLATED***foo</i></label>
			<label><input type="checkbox">***TRANSLATED*** checkbox <b>***TRANSLATED***bar</b></label>
			
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <i>***TRANSLATED***foo</i></label>
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <b>***TRANSLATED***bar</b></label>

			<select>
				<optgroup label="Numbers">
					<option value="1">***TRANSLATED***1</option>
					<option value="2">***TRANSLATED***2</option>
				</optgroup>
				<option value="foo">***TRANSLATED***foo</option>
				<option value="bar">***TRANSLATED***bar</option>
			</select>
		</form>

		<div class="custom-elements">
			<div class="notranslate">***TRANSLATED***This text must not be translated <span>***TRANSLATED***this and <b>***TRANSLATED***this</b>***TRANSLATED*** text too</span>***TRANSLATED***.</div>
			<div contenteditable="true">***TRANSLATED***This text must not be translated, since it is editable</div>

			<pre>***TRANSLATED***Text with preserved formatting</pre>

			<div translate="no">
				<div>***TRANSLATED***Items must not be translated:</div>
				<div>***TRANSLATED***Foo</div>
				<div>***TRANSLATED***Bar <span title="baz title">***TRANSLATED***baz</span></div>
			</div>

			<label><input type="checkbox" title="Item title" checked="">***TRANSLATED*** Checked item</label>
			<label><input type="checkbox" title="Item title">***TRANSLATED*** Unchecked item</label>
		</div>
	</div>
	<script>***TRANSLATED***console.log('console log: page end')</script>



</body></html>"
`;

exports[`basic usage > translate whole document with lazyTranslate 1`] = `
"<html lang="en"><head>
	<meta charset="UTF-8">
	<title>***TRANSLATED***Demo page with challenges for DOM translator</title>
	<meta name="viewport" content="width=device-width, initial-scale=1.0" data-rh="true">
	<link rel="search" type="application/opensearchdescription+xml" title="Opensearch" href="/opensearch.xml">
	<link rel="preload" href="#file.js" as="script">
	<script>***TRANSLATED***console.log('console log: page start')</script>
</head>

<body>***TRANSLATED***
	Text with no container
	<div id="rootElement" class="main container">***TRANSLATED***
		Text in div<br>
		<p>***TRANSLATED***Text inside container <a href="#url" title="link title">***TRANSLATED***link text</a></p>

		<img title="image" alt="image alt text" data-text="data attribute that must not be translated">
		<pre>***TRANSLATED***Text with preserved formatting</pre>***TRANSLATED***

		Some code:
		<pre>			<code lang="javascript">***TRANSLATED***
				const name = "Jeff";
				console.log("Your name is " + name);
			</code>
		</pre>

		<figure>
			<blockquote cite="https://www.huxley.net/bnw/four.html">
				<p>***TRANSLATED***Words can be like X-rays, if you use them properly—they’ll go through anything. You read and you’re pierced.</p>
			</blockquote>
			<figcaption>***TRANSLATED***—Aldous Huxley, <cite>***TRANSLATED***Brave New World</cite></figcaption>
		</figure>

		<form method="post" style="display: flex; flex-flow: column;">
			<input type="text" name="text" value="text">
			<input type="password" name="text" value="password">
			<textarea>***TRANSLATED***long text in textarea</textarea>

			<label><input type="checkbox">***TRANSLATED*** checkbox <i>***TRANSLATED***foo</i></label>
			<label><input type="checkbox">***TRANSLATED*** checkbox <b>***TRANSLATED***bar</b></label>
			
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <i>***TRANSLATED***foo</i></label>
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <b>***TRANSLATED***bar</b></label>

			<select>
				<optgroup label="Numbers">
					<option value="1">***TRANSLATED***1</option>
					<option value="2">***TRANSLATED***2</option>
				</optgroup>
				<option value="foo">***TRANSLATED***foo</option>
				<option value="bar">***TRANSLATED***bar</option>
			</select>
		</form>

		<div class="custom-elements">
			<div class="notranslate">***TRANSLATED***This text must not be translated <span>***TRANSLATED***this and <b>***TRANSLATED***this</b>***TRANSLATED*** text too</span>***TRANSLATED***.</div>
			<div contenteditable="true">***TRANSLATED***This text must not be translated, since it is editable</div>

			<pre>***TRANSLATED***Text with preserved formatting</pre>

			<div translate="no">
				<div>***TRANSLATED***Items must not be translated:</div>
				<div>***TRANSLATED***Foo</div>
				<div>***TRANSLATED***Bar <span title="baz title">***TRANSLATED***baz</span></div>
			</div>

			<label><input type="checkbox" title="Item title" checked="">***TRANSLATED*** Checked item</label>
			<label><input type="checkbox" title="Item title">***TRANSLATED*** Unchecked item</label>
		</div>
	</div>
	<script>console.log('console log: page end')</script>



</body></html>"
`;

exports[`translation with consider translatable and ignored nodes (lazy translation mode) > translate whole document 1`] = `
"<html lang="en"><head>
	<meta charset="UTF-8">
	<title>***TRANSLATED***Demo page with challenges for DOM translator</title>
	<meta name="viewport" content="width=device-width, initial-scale=1.0" data-rh="true">
	<link rel="search" type="application/opensearchdescription+xml" title="Opensearch" href="/opensearch.xml">
	<link rel="preload" href="#file.js" as="script">
	<script>console.log('console log: page start')</script>
</head>

<body>***TRANSLATED***
	Text with no container
	<div id="rootElement" class="main container">***TRANSLATED***
		Text in div<br>
		<p>***TRANSLATED***Text inside container <a href="#url" title="***TRANSLATED***link title">***TRANSLATED***link text</a></p>

		<img title="***TRANSLATED***image" alt="***TRANSLATED***image alt text" data-text="data attribute that must not be translated">
		<pre>***TRANSLATED***Text with preserved formatting</pre>***TRANSLATED***

		Some code:
		<pre>			<code lang="javascript">
				const name = "Jeff";
				console.log("Your name is " + name);
			</code>
		</pre>

		<figure>
			<blockquote cite="https://www.huxley.net/bnw/four.html">
				<p>***TRANSLATED***Words can be like X-rays, if you use them properly—they’ll go through anything. You read and you’re pierced.</p>
			</blockquote>
			<figcaption>***TRANSLATED***—Aldous Huxley, <cite>***TRANSLATED***Brave New World</cite></figcaption>
		</figure>

		<form method="post" style="display: flex; flex-flow: column;">
			<input type="text" name="text" value="text">
			<input type="password" name="text" value="password">
			<textarea>long text in textarea</textarea>

			<label><input type="checkbox">***TRANSLATED*** checkbox <i>***TRANSLATED***foo</i></label>
			<label><input type="checkbox">***TRANSLATED*** checkbox <b>***TRANSLATED***bar</b></label>
			
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <i>***TRANSLATED***foo</i></label>
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <b>***TRANSLATED***bar</b></label>

			<select>
				<optgroup label="***TRANSLATED***Numbers">
					<option value="1">***TRANSLATED***1</option>
					<option value="2">***TRANSLATED***2</option>
				</optgroup>
				<option value="foo">***TRANSLATED***foo</option>
				<option value="bar">***TRANSLATED***bar</option>
			</select>
		</form>

		<div class="custom-elements">
			<div class="notranslate">***TRANSLATED***This text must not be translated <span>***TRANSLATED***this and <b>***TRANSLATED***this</b>***TRANSLATED*** text too</span>***TRANSLATED***.</div>
			<div contenteditable="true">***TRANSLATED***This text must not be translated, since it is editable</div>

			<pre>***TRANSLATED***Text with preserved formatting</pre>

			<div translate="no">
				<div>***TRANSLATED***Items must not be translated:</div>
				<div>***TRANSLATED***Foo</div>
				<div>***TRANSLATED***Bar <span title="***TRANSLATED***baz title">***TRANSLATED***baz</span></div>
			</div>

			<label><input type="checkbox" title="***TRANSLATED***Item title" checked="">***TRANSLATED*** Checked item</label>
			<label><input type="checkbox" title="***TRANSLATED***Item title">***TRANSLATED*** Unchecked item</label>
		</div>
	</div>
	<script>console.log('console log: page end')</script>



</body></html>"
`;

exports[`translation with consider translatable and ignored nodes (lazy translation mode) > use custom nodes filter 1`] = `
"<html lang="en"><head>
	<meta charset="UTF-8">
	<title>***TRANSLATED***Demo page with challenges for DOM translator</title>
	<meta name="viewport" content="width=device-width, initial-scale=1.0" data-rh="true">
	<link rel="search" type="application/opensearchdescription+xml" title="Opensearch" href="/opensearch.xml">
	<link rel="preload" href="#file.js" as="script">
	<script>console.log('console log: page start')</script>
</head>

<body>***TRANSLATED***
	Text with no container
	<div id="rootElement" class="main container">***TRANSLATED***
		Text in div<br>
		<p>***TRANSLATED***Text inside container <a href="#url" title="***TRANSLATED***link title">***TRANSLATED***link text</a></p>

		<img title="***TRANSLATED***image" alt="***TRANSLATED***image alt text" data-text="data attribute that must not be translated">
		<pre>***TRANSLATED***Text with preserved formatting</pre>***TRANSLATED***

		Some code:
		<pre>			<code lang="javascript">
				const name = "Jeff";
				console.log("Your name is " + name);
			</code>
		</pre>

		<figure>
			<blockquote cite="https://www.huxley.net/bnw/four.html">
				<p>***TRANSLATED***Words can be like X-rays, if you use them properly—they’ll go through anything. You read and you’re pierced.</p>
			</blockquote>
			<figcaption>***TRANSLATED***—Aldous Huxley, <cite>***TRANSLATED***Brave New World</cite></figcaption>
		</figure>

		<form method="post" style="display: flex; flex-flow: column;">
			<input type="text" name="text" value="text">
			<input type="password" name="text" value="password">
			<textarea>long text in textarea</textarea>

			<label><input type="checkbox">***TRANSLATED*** checkbox <i>***TRANSLATED***foo</i></label>
			<label><input type="checkbox">***TRANSLATED*** checkbox <b>***TRANSLATED***bar</b></label>
			
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <i>***TRANSLATED***foo</i></label>
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <b>***TRANSLATED***bar</b></label>

			<select>
				<optgroup label="***TRANSLATED***Numbers">
					<option value="1">***TRANSLATED***1</option>
					<option value="2">***TRANSLATED***2</option>
				</optgroup>
				<option value="foo">***TRANSLATED***foo</option>
				<option value="bar">***TRANSLATED***bar</option>
			</select>
		</form>

		<div class="custom-elements">
			<div class="notranslate">This text must not be translated <span>this and <b>this</b> text too</span>.</div>
			<div contenteditable="true">This text must not be translated, since it is editable</div>

			<pre>***TRANSLATED***Text with preserved formatting</pre>

			<div translate="no">
				<div>Items must not be translated:</div>
				<div>Foo</div>
				<div>Bar <span title="baz title">baz</span></div>
			</div>

			<label><input type="checkbox" title="Item title" checked="">***TRANSLATED*** Checked item</label>
			<label><input type="checkbox" title="***TRANSLATED***Item title">***TRANSLATED*** Unchecked item</label>
		</div>
	</div>
	<script>console.log('console log: page end')</script>



</body></html>"
`;

exports[`translation with consider translatable and ignored nodes > translate whole document 1`] = `
"<html lang="en"><head>
	<meta charset="UTF-8">
	<title>***TRANSLATED***Demo page with challenges for DOM translator</title>
	<meta name="viewport" content="width=device-width, initial-scale=1.0" data-rh="true">
	<link rel="search" type="application/opensearchdescription+xml" title="Opensearch" href="/opensearch.xml">
	<link rel="preload" href="#file.js" as="script">
	<script>console.log('console log: page start')</script>
</head>

<body>***TRANSLATED***
	Text with no container
	<div id="rootElement" class="main container">***TRANSLATED***
		Text in div<br>
		<p>***TRANSLATED***Text inside container <a href="#url" title="***TRANSLATED***link title">***TRANSLATED***link text</a></p>

		<img title="***TRANSLATED***image" alt="***TRANSLATED***image alt text" data-text="data attribute that must not be translated">
		<pre>***TRANSLATED***Text with preserved formatting</pre>***TRANSLATED***

		Some code:
		<pre>			<code lang="javascript">
				const name = "Jeff";
				console.log("Your name is " + name);
			</code>
		</pre>

		<figure>
			<blockquote cite="https://www.huxley.net/bnw/four.html">
				<p>***TRANSLATED***Words can be like X-rays, if you use them properly—they’ll go through anything. You read and you’re pierced.</p>
			</blockquote>
			<figcaption>***TRANSLATED***—Aldous Huxley, <cite>***TRANSLATED***Brave New World</cite></figcaption>
		</figure>

		<form method="post" style="display: flex; flex-flow: column;">
			<input type="text" name="text" value="text">
			<input type="password" name="text" value="password">
			<textarea>long text in textarea</textarea>

			<label><input type="checkbox">***TRANSLATED*** checkbox <i>***TRANSLATED***foo</i></label>
			<label><input type="checkbox">***TRANSLATED*** checkbox <b>***TRANSLATED***bar</b></label>
			
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <i>***TRANSLATED***foo</i></label>
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <b>***TRANSLATED***bar</b></label>

			<select>
				<optgroup label="***TRANSLATED***Numbers">
					<option value="1">***TRANSLATED***1</option>
					<option value="2">***TRANSLATED***2</option>
				</optgroup>
				<option value="foo">***TRANSLATED***foo</option>
				<option value="bar">***TRANSLATED***bar</option>
			</select>
		</form>

		<div class="custom-elements">
			<div class="notranslate">***TRANSLATED***This text must not be translated <span>***TRANSLATED***this and <b>***TRANSLATED***this</b>***TRANSLATED*** text too</span>***TRANSLATED***.</div>
			<div contenteditable="true">***TRANSLATED***This text must not be translated, since it is editable</div>

			<pre>***TRANSLATED***Text with preserved formatting</pre>

			<div translate="no">
				<div>***TRANSLATED***Items must not be translated:</div>
				<div>***TRANSLATED***Foo</div>
				<div>***TRANSLATED***Bar <span title="***TRANSLATED***baz title">***TRANSLATED***baz</span></div>
			</div>

			<label><input type="checkbox" title="***TRANSLATED***Item title" checked="">***TRANSLATED*** Checked item</label>
			<label><input type="checkbox" title="***TRANSLATED***Item title">***TRANSLATED*** Unchecked item</label>
		</div>
	</div>
	<script>console.log('console log: page end')</script>



</body></html>"
`;

exports[`translation with consider translatable and ignored nodes > use custom nodes filter 1`] = `
"<html lang="en"><head>
	<meta charset="UTF-8">
	<title>***TRANSLATED***Demo page with challenges for DOM translator</title>
	<meta name="viewport" content="width=device-width, initial-scale=1.0" data-rh="true">
	<link rel="search" type="application/opensearchdescription+xml" title="Opensearch" href="/opensearch.xml">
	<link rel="preload" href="#file.js" as="script">
	<script>console.log('console log: page start')</script>
</head>

<body>***TRANSLATED***
	Text with no container
	<div id="rootElement" class="main container">***TRANSLATED***
		Text in div<br>
		<p>***TRANSLATED***Text inside container <a href="#url" title="***TRANSLATED***link title">***TRANSLATED***link text</a></p>

		<img title="***TRANSLATED***image" alt="***TRANSLATED***image alt text" data-text="data attribute that must not be translated">
		<pre>***TRANSLATED***Text with preserved formatting</pre>***TRANSLATED***

		Some code:
		<pre>			<code lang="javascript">
				const name = "Jeff";
				console.log("Your name is " + name);
			</code>
		</pre>

		<figure>
			<blockquote cite="https://www.huxley.net/bnw/four.html">
				<p>***TRANSLATED***Words can be like X-rays, if you use them properly—they’ll go through anything. You read and you’re pierced.</p>
			</blockquote>
			<figcaption>***TRANSLATED***—Aldous Huxley, <cite>***TRANSLATED***Brave New World</cite></figcaption>
		</figure>

		<form method="post" style="display: flex; flex-flow: column;">
			<input type="text" name="text" value="text">
			<input type="password" name="text" value="password">
			<textarea>long text in textarea</textarea>

			<label><input type="checkbox">***TRANSLATED*** checkbox <i>***TRANSLATED***foo</i></label>
			<label><input type="checkbox">***TRANSLATED*** checkbox <b>***TRANSLATED***bar</b></label>
			
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <i>***TRANSLATED***foo</i></label>
			<label><input type="radio" name="radioGroup">***TRANSLATED*** radio <b>***TRANSLATED***bar</b></label>

			<select>
				<optgroup label="***TRANSLATED***Numbers">
					<option value="1">***TRANSLATED***1</option>
					<option value="2">***TRANSLATED***2</option>
				</optgroup>
				<option value="foo">***TRANSLATED***foo</option>
				<option value="bar">***TRANSLATED***bar</option>
			</select>
		</form>

		<div class="custom-elements">
			<div class="notranslate">This text must not be translated <span>this and <b>this</b> text too</span>.</div>
			<div contenteditable="true">This text must not be translated, since it is editable</div>

			<pre>***TRANSLATED***Text with preserved formatting</pre>

			<div translate="no">
				<div>Items must not be translated:</div>
				<div>Foo</div>
				<div>Bar <span title="baz title">baz</span></div>
			</div>

			<label><input type="checkbox" title="Item title" checked="">***TRANSLATED*** Checked item</label>
			<label><input type="checkbox" title="***TRANSLATED***Item title">***TRANSLATED*** Unchecked item</label>
		</div>
	</div>
	<script>console.log('console log: page end')</script>



</body></html>"
`;
